File: src/main/java/com/yasarbilgi/announcementtracker/service/AnnouncementService.java
package com.yasarbilgi.announcementtracker.service;

import com.yasarbilgi.announcementtracker.dto.response.AnnouncementResponseDto;
import com.yasarbilgi.announcementtracker.enums.SiteType;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

import java.util.List;
import java.util.Map;
import java.util.Set;

public interface AnnouncementService {

    /**
     * Triggers scraping for all registered site strategies.
     * @return List of newly added announcements.
     */
    List<AnnouncementResponseDto> triggerScrapeAll();

    /**
     * Triggers scraping for a specific site strategy.
     * @return List of newly added announcements for that site.
     */
    List<AnnouncementResponseDto> triggerScrapeSite(SiteType siteType);

    /**
     * Fetches stored announcements with pagination.
     */
    Page<AnnouncementResponseDto> getAllAnnouncements(
            SiteType siteType, String search, boolean hasAttachment, Pageable pageable);

    /**
     * Fetches stored announcements for a user's subscribed sites with site filtering and pagination.
     */
    Page<AnnouncementResponseDto> getAnnouncementsForSites(
            Set<SiteType> subscribedSites,
            SiteType siteFilter,
            String search,
            boolean hasAttachment,
            Pageable pageable);

    /**
     * Returns the persisted announcement total for every source.
     */
    Map<SiteType, Long> getAnnouncementCounts();

    /**
     * Retrieves announcement by ID.
     */
    AnnouncementResponseDto getAnnouncementById(Long id);

    /**
     * Dispatches pending un-notified announcements to subscribers.
     */
    int notifyPendingAnnouncements();

    /**
     * Sends a instant test email with the latest announcement to active subscribers.
     */
    void sendTestEmail();
}

File: src/main/resources/static/js/services/announcementService.js
/**
 * Announcement Service
 */
window.AnnouncementService = {
    getAnnouncements: async function(page = 0, size = 10, site = '', search = '', hasAttachment = false) {
        let url = `/api/v1/announcements?page=${page}&size=${size}`;
        if (site) url += `&siteType=${encodeURIComponent(site)}`;
        if (search) url += `&search=${encodeURIComponent(search)}`;
        if (hasAttachment) url += '&hasAttachment=true';
        const res = await window.ApiClient.get(url);
        return res.data || res;
    },

    getUserAnnouncements: async function(page = 0, size = 10, site = '', search = '', hasAttachment = false) {
        let url = `/api/v1/user/me/announcements?page=${page}&size=${size}`;
        if (site) url += `&siteType=${encodeURIComponent(site)}`;
        if (search) url += `&search=${encodeURIComponent(search)}`;
        if (hasAttachment) url += '&hasAttachment=true';
        const res = await window.ApiClient.get(url);
        return res.data || res;
    },

    getAnnouncementCounts: async function() {
        const res = await window.ApiClient.get('/api/v1/announcements/counts');
        return res.data || res;
    },

    triggerScrape: async function() {
        const res = await window.ApiClient.post('/api/v1/announcements/trigger');
        return res.data || res;
    },

    getSites: async function() {
        const res = await window.ApiClient.get('/api/v1/sites');
        return res.data || res;
    }
};
